requirements/0003-audit-trail.md

```diff
@@ -0,0 +1,89 @@
+- SafeShelter Issue: https://safeshelter.youtrack.cloud/issue/RFC-2
+
+# Summary
+
+This is an RFC to define audit trail for the system.
+
+# Motivation
+
+For audit and transparency into the process purposes, the system cannot just overwrite some data. Organization Admins need to be able to see the history of *Guest* assignments to *Accommodation Units* or which *Team Member* worked on which *Guest* *Entry*.
+
+# Detailed Design
+
+## Use of IETF Keywords
+
+This document employs a subset of the Internet Engineering Task Force keywords found in [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119). These words are MUST, SHALL, SHOULD, MAY and their counterparts MUST NOT, SHOULD NOT, MAY NOT. They are capitalized throughout the document to draw attention to their special status as keywords used to indicate requirements levels.
+
+Readers are directed to interpret them as requirements at levels consistent with their term of art definitions in the IETF [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119).
+
+## Business Architecture
+
+### Capabilities
+
+Provide a way to track important state changes to the application's data, without the need to parse the logs.
+
+### Assumptions
+
+1. Functionality SHALL not be developed from scratch. An existing solution SHALL be used.
+
+### Dependencies
+
+There are available off-the-shelf FOSS solutions out there:
+- [pgMemento](https://github.com/pgMemento/pgMemento) - it is bare PostgreSQL solution using triggers. It is able to track  DDL changes, but it is not necessary and maybe even problematic given Alembic migrations. Stores audit data as JSONB deltas.
+- [SQLAlchemy-Continuum](https://github.com/kvesteri/SQLAlchemy-Continuum) - A proper SQLAlchemy extentions. Actively maintained. Supports Alembic migrations. Does not seem to support tracking WHO made a change. 78k downloads a month.
+- [Audit Trigger 91plus](https://github.com/2ndQuadrant/audit-trigger) - Does not track `SELECT` or `DDL` changes, but it is not required. Not sure how it'll work with Alembic migrations.
+- [PostgreSQL-Audit](https://github.com/kvesteri/postgresql-audit) - Seems like a next version of SQLAlchemy-Continuum. README claims it, and it's by the same developer. It has tho less downloads (6k vs 78k), less stars, less commits, and last change was done almost a year ago, compared to month ago in Continuum.
+
+Recommendation is to go with [SQLAlchemy-Continuum](https://github.com/kvesteri/SQLAlchemy-Continuum). Since it -- nor apparently neither of the plugins -- offers functionality of logging *who* made a change, development of custom plugin MAY be necessary.
+
+
+### Risks 
+
+### Business Risks
+
+Storing too much data in the separate audit table may create data breach issues, as well as make RODO-related activities like *right to be forgotten* more complex.
+
+### Impacts
+
+None.
+
+## Data Architecture
+
+A new table SHALL be created, tracking at least:
```

Comment:
The reality of `SQLAlchemy-Continuum` is as follows:
* there is a `transaction` table with columns `issued_at` and `id`
* for actual versions data we have a table per entity type, not one common table
* each versions table has the same schema as the normal entities table + following additional columns:
  * `transaction_id`, fk to the `transaction` table
  * `end_transaction_id`, the id of the next transaction that touched the same row, `null` if the current row is the latest version
  * `operation_type` – CREATE / UPDATE / DELETE, coded in a `smallint`
* values in the row are from *after* the change (not sure how it works for deletes tho)
* ^ this means that there is no need to extend continuum to track, e.g., who made the change. Instead, we can just add a `updated_by` column to each app table, and then we'll get the data we need.

Reply:
This is probably too detailed for an RFC, I just wanted to sell you all the current context to decide how/if to adjust the RFC